### PR TITLE
Move at::{Refcounted,}MapAllocator to c10

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109881

`libshm.so` depends on the torch library exclusively for `at::RefcountedMapAllocator`,
 so it makes sense to move it to c10 along with the other memory allocators.

This means `libshm.so` only depends on `c10` and we don't need to relink
`libshm.so` for every ATen change.

### CODE REVIEW

File: c10/core/MapAllocator.h [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  ALLOCATOR_MAPPED_UNLINK = 64
+};
+
+// Sentinel value/type to help distinguish the file descriptor constructor from
+// the non-file descriptor constructor
+enum WithFd { WITH_FD };
+
+C10_API std::string NewProcessWideShmHandle();
+
+class C10_API MapAllocator {
```
Comment (peterbell10) [inline c10/core/MapAllocator.h:25]: These diffs are a mess, but it's mostly just a copy paste of the aten files with `aten` changed to `c10`. However, with the linter changes, it's too big of a diff for git to recognize as a simple rename. This is clearer in the commit-level diffs.

**Review by albanD — APPROVED** [review]
Ho very nice cleanup!

I'll verify that all build are good and then merge this.

**Comment by albanD** [comment]
I can't import this so yolo

@pytorchbot merge -r

File: c10/core/build.bzl [hunk 2]
```diff
@@ -18,6 +18,27 @@ def define_targets(rules):
...
+            "HAVE_SHM_OPEN=1",
+            "HAVE_SHM_UNLINK=1",
+        ],
+        visibility = ["//visibility:public"],
+        deps = [
+            ":CPUAllocator",
+            ":base",
+            "//c10/util:base",
+        ],
+        linkopts = ["-lrt"],
```
Comment (albanD) [inline c10/core/build.bzl:38]: This is not actually a valid argument for buck rules.
Also this is not portable and rt should be pulled in as a deps.

**Comment by albanD** [comment]
Is the ios failure relevant? The others are not for sure.

**Comment by peterbell10** [comment]
iOS tests seems to be passing after the rebase so probably just a flake.

**Comment by albanD** [comment]
Ok!
Feel free to merge this again once CI is green!

**Comment by kit1980** [comment]
@peterbell10 You probably need someone from Meta to help re-landing this.

File: c10/BUCK.oss [hunk 3]
```diff
@@ -44,7 +45,12 @@ cxx_library(
     header_namespace = "c10",
     link_whole = True,
     platform_preprocessor_flags = [['windows', ['-D_WINDOWS', '-D_WIN32', '-DWIN32', '-DNOMINMAX', '-D_CRT_SECURE_NO_WARNINGS', '-D_USE_MATH_DEFINES']], ['windows.*64$', ['-D_WIN64']]],
-    preprocessor_flags = ['-DC10_BUILD_MAIN_LIB'],
+    preprocessor_flags = [
+        '-DC10_BUILD_MAIN_LIB',
+        '-DHAVE_MMAP=1',
```
Comment (peterbell10) [inline c10/BUCK.oss:50]:
```
undefined symbol: _ZN3c1022RefcountedMapAllocator6decrefEv"
```

I had the same error in the bazel build when `HAVE_MMAP` wasn't being defined. Is it possible there is another build definition that needs the flag added?

If so, does that mean we have 4 completely unrelated build definitions: CMake, bazel, buck (oss), buck (internal)?

Comment (kit1980) [reply c10/BUCK.oss:50]: @peterbell10 unfortunately it's even more than 4: more than 1 internal buck.

But they are not completely unrelated: all of them are supposed to use build.bzl files (like https://github.com/pytorch/pytorch/blob/main/c10/build.bzl), but it's only implemented for some things so far.

For this particular case, likely https://github.com/pytorch/pytorch/blob/main/c10/ovrsource_defs.bzl needs to be updated. But someone from Meta should work with you on this.

File: c10/ovrsource_defs.bzl [hunk 4]
```diff
@@ -65,7 +68,9 @@ def define_c10_ovrsource(name, is_mobile):
         visibility = [
             "//xplat/caffe2/c10:c10_ovrsource",
         ],
-        deps = select({
+        deps = [
+            "//third-party/rt",
+        ] + select({
```
Comment (peterbell10) [inline c10/ovrsource_defs.bzl:73]: @kit1980 would you be able to help on the meta-internal side? I think this might fix it but not idea if it works.

Comment (kit1980) [reply c10/ovrsource_defs.bzl:73]: I can try but not today.

Comment (peterbell10) [reply c10/ovrsource_defs.bzl:73]: @kit1980 any updates on this?

Comment (lezcano) [reply c10/ovrsource_defs.bzl:73]: ping @kit1980

Comment (kit1980) [reply c10/ovrsource_defs.bzl:73]: I'll look today.

**Comment by kit1980** [comment]
@peterbell10 May I ask you to fix the conflicts?
Or even better create a new non-ghstack PR - it will simplify things for me (it's a single PR stack anyway).

**Comment by peterbell10** [comment]
Okay, rebased and resubmitted as #116204

**Pull request closed without merging**